File: implementations/csharp/client/src/Specus.Client/PeerMesh/NatBehaviorDiscovery.cs
using System.Net;

namespace Specus.Client.PeerMesh;

internal static class NatBehavior
{
    public const string DiscoveryRfc5780 = "RFC5780";
    public const string DiscoveryBasic = "BASIC";

    public const string EndpointIndependent = "ENDPOINT_INDEPENDENT";
    public const string AddressDependent = "ADDRESS_DEPENDENT";
    public const string AddressAndPortDependent = "ADDRESS_AND_PORT_DEPENDENT";
    public const string Unknown = "UNKNOWN";
    public const string Unsupported = "UNSUPPORTED";
}

internal enum NatBehaviorProbe
{
    FilterChangeIpAndPort,
    FilterChangePort,
    MappingAlternateIp,
    MappingAlternateIpAndPort,
}

internal sealed record NatBehaviorProbeRequest(
    int Generation,
    NatBehaviorProbe Probe,
    IPEndPoint TargetEndpoint,
    IPEndPoint ExpectedResponseEndpoint,
    bool ChangeIp,
    bool ChangePort);

internal sealed record NatBehaviorSnapshot(
    int Generation,
    string Discovery,
    string MappingBehavior,
    string FilteringBehavior,
    IPEndPoint? MappedEndpoint,
    bool Complete);

internal sealed record NatBehaviorTransition(
    bool Accepted,
    NatBehaviorProbeRequest? NextProbe,
    NatBehaviorSnapshot Snapshot);

internal sealed class NatBehaviorDiscovery
{
    private readonly object _sync = new();
    private int _generation;
    private NatBehaviorProbe? _pendingProbe;
    private IPEndPoint? _primaryEndpoint;
    private IPEndPoint? _otherEndpoint;
    private IPEndPoint? _primaryMappedEndpoint;
    private IPEndPoint? _alternateIpMappedEndpoint;
    private string _mappingBehavior = "";
    private string _filteringBehavior = "";
    private bool _complete;

    public NatBehaviorTransition Begin(
        IPEndPoint primaryEndpoint,
        IPEndPoint primaryMappedEndpoint,
        IPEndPoint otherEndpoint)
    {
        lock (_sync)
        {
            RequireEndpoint(primaryEndpoint, nameof(primaryEndpoint));
            RequireEndpoint(primaryMappedEndpoint, nameof(primaryMappedEndpoint));
            RequireEndpoint(otherEndpoint, nameof(otherEndpoint));
            if (primaryEndpoint.Address.Equals(otherEndpoint.Address)
                || primaryEndpoint.Port == otherEndpoint.Port)
            {
                throw new ArgumentException(
                    "RFC 5780 discovery requires another IP address and another UDP port");
            }

            _generation++;
            _primaryEndpoint = Clone(primaryEndpoint);
            _primaryMappedEndpoint = Clone(primaryMappedEndpoint);
            _otherEndpoint = Clone(otherEndpoint);
            _alternateIpMappedEndpoint = null;
            _mappingBehavior = "";
            _filteringBehavior = "";
            _complete = false;
            return Next(NatBehaviorProbe.FilterChangeIpAndPort);
        }
    }

    public NatBehaviorTransition Succeeded(
        int expectedGeneration,
        NatBehaviorProbe probe,
        IPEndPoint mappedEndpoint)
    {
        lock (_sync)
        {
            if (!Accepts(expectedGeneration, probe))
            {
                return Ignored();
            }
            RequireEndpoint(mappedEndpoint, nameof(mappedEndpoint));
            return probe switch
            {
                NatBehaviorProbe.FilterChangeIpAndPort => SetFilteringAndNext(
                    NatBehavior.EndpointIndependent,
                    NatBehaviorProbe.MappingAlternateIp),
                NatBehaviorProbe.FilterChangePort => SetFilteringAndNext(
                    NatBehavior.AddressDependent,
                    NatBehaviorProbe.MappingAlternateIp),
                NatBehaviorProbe.MappingAlternateIp => SetAlternateMappingAndNext(mappedEndpoint),
                NatBehaviorProbe.MappingAlternateIpAndPort => FinishMapping(mappedEndpoint),
                _ => Ignored(),
            };
        }
    }

    public NatBehaviorTransition TimedOut(int expectedGeneration, NatBehaviorProbe probe)
    {
        lock (_sync)
        {
            if (!Accepts(expectedGeneration, probe))
            {
                return Ignored();
            }
            return probe switch
            {
                NatBehaviorProbe.FilterChangeIpAndPort => Next(NatBehaviorProbe.FilterChangePort),
                NatBehaviorProbe.FilterChangePort => SetFilteringAndNext(
                    NatBehavior.AddressAndPortDependent,
                    NatBehaviorProbe.MappingAlternateIp),
                NatBehaviorProbe.MappingAlternateIp or NatBehaviorProbe.MappingAlternateIpAndPort =>
                    FinishMappingFailure(),
                _ => Ignored(),
            };
        }
    }

    public NatBehaviorTransition Failed(
        int expectedGeneration,
        NatBehaviorProbe probe,
        bool unsupported)
    {
        lock (_sync)
        {
            if (!Accepts(expectedGeneration, probe))
            {
                return Ignored();
            }
            return probe switch
            {
                NatBehaviorProbe.FilterChangeIpAndPort or NatBehaviorProbe.FilterChangePort =>
                    SetFilteringAndNext(
                        unsupported ? NatBehavior.Unsupported : NatBehavior.Unknown,
                        NatBehaviorProbe.MappingAlternateIp),
                NatBehaviorProbe.MappingAlternateIp or NatBehaviorProbe.MappingAlternateIpAndPort =>
                    FinishMappingFailure(),
                _ => Ignored(),
            };
        }
    }

    public NatBehaviorSnapshot Snapshot()
    {
        lock (_sync)
        {
            return SnapshotLocked();
        }
    }

    private NatBehaviorTransition SetFilteringAndNext(string filtering, NatBehaviorProbe next)
    {
        _filteringBehavior = filtering;
        return Next(next);
    }

    private NatBehaviorTransition SetAlternateMappingAndNext(IPEndPoint mappedEndpoint)
    {
        _alternateIpMappedEndpoint = Clone(mappedEndpoint);
        return Next(NatBehaviorProbe.MappingAlternateIpAndPort);
    }

    private NatBehaviorTransition FinishMapping(IPEndPoint alternateIpAndPortMappedEndpoint)
    {
        _mappingBehavior = SameEndpoint(_primaryMappedEndpoint, _alternateIpMappedEndpoint)
            ? NatBehavior.EndpointIndependent
            : SameEndpoint(_alternateIpMappedEndpoint, alternateIpAndPortMappedEndpoint)
                ? NatBehavior.AddressDependent
                : NatBehavior.AddressAndPortDependent;
        _pendingProbe = null;
        _complete = true;
        return Accepted(null);
    }

    private NatBehaviorTransition FinishMappingFailure()
    {
        _mappingBehavior = NatBehavior.Unknown;
        if (!string.Equals(_filteringBehavior, NatBehavior.EndpointIndependent, StringComparison.Ordinal)
            && !string.Equals(_filteringBehavior, NatBehavior.Unsupported, StringComparison.Ordinal))
        {
            _filteringBehavior = NatBehavior.Unknown;
        }
        _pendingProbe = null;
        _complete = true;
        return Accepted(null);
    }

    private NatBehaviorTransition Next(NatBehaviorProbe probe)
    {
        _pendingProbe = probe;
        return Accepted(Request(probe));
    }

    private NatBehaviorProbeRequest Request(NatBehaviorProbe probe)
    {
        var primary = _primaryEndpoint!;
        var other = _otherEndpoint!;
        var alternateIpPrimaryPort = new IPEndPoint(other.Address, primary.Port);
        var primaryIpAlternatePort = new IPEndPoint(primary.Address, other.Port);
        return probe switch
        {
            NatBehaviorProbe.FilterChangeIpAndPort => new(
                _generation,
                probe,
                Clone(primary),
                Clone(other),
                true,
                true),
            NatBehaviorProbe.FilterChangePort => new(
                _generation,
                probe,
                Clone(primary),
                primaryIpAlternatePort,
                false,
                true),
            NatBehaviorProbe.MappingAlternateIp => new(
                _generation,
                probe,
                alternateIpPrimaryPort,
                Clone(alternateIpPrimaryPort),
                false,
                false),
            NatBehaviorProbe.MappingAlternateIpAndPort => new(
                _generation,
                probe,
                Clone(other),
                Clone(other),
                false,
                false),
            _ => throw new ArgumentOutOfRangeException(nameof(probe)),
        };
    }

    private bool Accepts(int expectedGeneration, NatBehaviorProbe probe) =>
        expectedGeneration == _generation && _pendingProbe == probe && !_complete;

    private NatBehaviorTransition Accepted(NatBehaviorProbeRequest? nextProbe) =>
        new(true, nextProbe, SnapshotLocked());

    private NatBehaviorTransition Ignored() =>
        new(false, null, SnapshotLocked());

    private NatBehaviorSnapshot SnapshotLocked() =>
        new(
            _generation,
            NatBehavior.DiscoveryRfc5780,
            _mappingBehavior,
            _filteringBehavior,
            _primaryMappedEndpoint is null ? null : Clone(_primaryMappedEndpoint),
            _complete);

    private static void RequireEndpoint(IPEndPoint? endpoint, string name)
    {
        if (endpoint is null || endpoint.Port <= 0 || IsUnspecified(endpoint.Address))
        {
            throw new ArgumentException($"{name} must be a resolved UDP endpoint", name);
        }
    }

    private static bool IsUnspecified(IPAddress address) =>
        address.Equals(IPAddress.Any) || address.Equals(IPAddress.IPv6Any);

    private static bool SameEndpoint(IPEndPoint? first, IPEndPoint? second) =>
        first is not null
        && second is not null
        && first.Port == second.Port
        && first.Address.Equals(second.Address);

    private static IPEndPoint Clone(IPEndPoint endpoint) =>
        new(new IPAddress(endpoint.Address.GetAddressBytes()), endpoint.Port);
}

File: implementations/java/client/src/main/java/com/theshuai/specusclient/peer/NatBehaviorDiscovery.java
package com.theshuai.specusclient.peer;

import java.net.InetSocketAddress;
import java.util.Objects;

/**
 * RFC 5780 mapping/filtering discovery state. Network I/O and retransmission
 * stay in {@link PeerMeshClient}; this class only owns probe ordering and
 * classification.
 */
final class NatBehaviorDiscovery {
    static final String DISCOVERY_RFC5780 = "RFC5780";
    static final String DISCOVERY_BASIC = "BASIC";

    static final String ENDPOINT_INDEPENDENT = "ENDPOINT_INDEPENDENT";
    static final String ADDRESS_DEPENDENT = "ADDRESS_DEPENDENT";
    static final String ADDRESS_AND_PORT_DEPENDENT = "ADDRESS_AND_PORT_DEPENDENT";
    static final String UNKNOWN = "UNKNOWN";
    static final String UNSUPPORTED = "UNSUPPORTED";

    private int generation;
    private Probe pendingProbe;
    private InetSocketAddress primaryEndpoint;
    private InetSocketAddress otherEndpoint;
    private InetSocketAddress primaryMappedEndpoint;
    private InetSocketAddress alternateIpMappedEndpoint;
    private String mappingBehavior = "";
    private String filteringBehavior = "";
    private boolean complete;

    synchronized Transition begin(InetSocketAddress primaryEndpoint,
                                  InetSocketAddress primaryMappedEndpoint,
                                  InetSocketAddress otherEndpoint) {
        requireResolved(primaryEndpoint, "primaryEndpoint");
        requireResolved(primaryMappedEndpoint, "primaryMappedEndpoint");
        requireResolved(otherEndpoint, "otherEndpoint");
        if (sameAddress(primaryEndpoint, otherEndpoint)
                || primaryEndpoint.getPort() == otherEndpoint.getPort()) {
            throw new IllegalArgumentException(
                    "RFC 5780 discovery requires another IP address and another UDP port");
        }

        generation++;
        this.primaryEndpoint = primaryEndpoint;
        this.primaryMappedEndpoint = primaryMappedEndpoint;
        this.otherEndpoint = otherEndpoint;
        alternateIpMappedEndpoint = null;
        mappingBehavior = "";
        filteringBehavior = "";
        complete = false;
        return next(Probe.FILTER_CHANGE_IP_AND_PORT);
    }

    synchronized Transition succeeded(int expectedGeneration,
                                      Probe probe,
                                      InetSocketAddress mappedEndpoint) {
        if (!accepts(expectedGeneration, probe)) {
            return ignored();
        }
        requireResolved(mappedEndpoint, "mappedEndpoint");
        return switch (probe) {
            case FILTER_CHANGE_IP_AND_PORT -> {
                filteringBehavior = ENDPOINT_INDEPENDENT;
                yield next(Probe.MAPPING_ALTERNATE_IP);
            }
            case FILTER_CHANGE_PORT -> {
                filteringBehavior = ADDRESS_DEPENDENT;
                yield next(Probe.MAPPING_ALTERNATE_IP);
            }
            case MAPPING_ALTERNATE_IP -> {
                alternateIpMappedEndpoint = mappedEndpoint;
                // Test III also proves A2:P2 is operational. That validation is
                // needed before treating a missing filtering response as NAT behavior.
                yield next(Probe.MAPPING_ALTERNATE_IP_AND_PORT);
            }
            case MAPPING_ALTERNATE_IP_AND_PORT -> finishMapping(mappedEndpoint);
        };
    }

    synchronized Transition timedOut(int expectedGeneration, Probe probe) {
        if (!accepts(expectedGeneration, probe)) {
            return ignored();
        }
        return switch (probe) {
            case FILTER_CHANGE_IP_AND_PORT -> next(Probe.FILTER_CHANGE_PORT);
            case FILTER_CHANGE_PORT -> {
                filteringBehavior = ADDRESS_AND_PORT_DEPENDENT;
                yield next(Probe.MAPPING_ALTERNATE_IP);
            }
            case MAPPING_ALTERNATE_IP, MAPPING_ALTERNATE_IP_AND_PORT -> finishMappingFailure();
        };
    }

    synchronized Transition failed(int expectedGeneration, Probe probe, boolean unsupported) {
        if (!accepts(expectedGeneration, probe)) {
            return ignored();
        }
        return switch (probe) {
            case FILTER_CHANGE_IP_AND_PORT, FILTER_CHANGE_PORT -> {
                filteringBehavior = unsupported ? UNSUPPORTED : UNKNOWN;
                yield next(Probe.MAPPING_ALTERNATE_IP);
            }
            case MAPPING_ALTERNATE_IP, MAPPING_ALTERNATE_IP_AND_PORT -> finishMappingFailure();
        };
    }

    synchronized Snapshot snapshot() {
        return new Snapshot(
                generation,
                DISCOVERY_RFC5780,
                mappingBehavior,
                filteringBehavior,
                primaryMappedEndpoint,
                complete);
    }

    synchronized boolean isActive() {
        return pendingProbe != null && !complete;
    }

    private Transition finishMapping(InetSocketAddress alternateIpAndPortMappedEndpoint) {
        if (sameEndpoint(primaryMappedEndpoint, alternateIpMappedEndpoint)) {
            mappingBehavior = ENDPOINT_INDEPENDENT;
        } else if (sameEndpoint(alternateIpMappedEndpoint, alternateIpAndPortMappedEndpoint)) {
            mappingBehavior = ADDRESS_DEPENDENT;
        } else {
            mappingBehavior = ADDRESS_AND_PORT_DEPENDENT;
        }
        pendingProbe = null;
        complete = true;
        return accepted(null);
    }

    private Transition finishMappingFailure() {
        mappingBehavior = UNKNOWN;
        // Endpoint-independent filtering was confirmed by an actual A2:P2
        // response. Other filtering outcomes depend on missing A2 traffic and
        // cannot be trusted when the normal A2 validation probe also failed.
        if (!ENDPOINT_INDEPENDENT.equals(filteringBehavior)
                && !UNSUPPORTED.equals(filteringBehavior)) {
            filteringBehavior = UNKNOWN;
        }
        pendingProbe = null;
        complete = true;
        return accepted(null);
    }

    private Transition next(Probe probe) {
        pendingProbe = probe;
        return accepted(request(probe));
    }

    private ProbeRequest request(Probe probe) {
        InetSocketAddress alternateIpPrimaryPort =
                new InetSocketAddress(otherEndpoint.getAddress(), primaryEndpoint.getPort());
        InetSocketAddress primaryIpAlternatePort =
                new InetSocketAddress(primaryEndpoint.getAddress(), otherEndpoint.getPort());
        return switch (probe) {
            case FILTER_CHANGE_IP_AND_PORT -> new ProbeRequest(
                    generation,
                    probe,
                    primaryEndpoint,
                    otherEndpoint,
                    true,
                    true);
            case FILTER_CHANGE_PORT -> new ProbeRequest(
                    generation,
                    probe,
                    primaryEndpoint,
                    primaryIpAlternatePort,
                    false,
                    true);
            case MAPPING_ALTERNATE_IP -> new ProbeRequest(
                    generation,
                    probe,
                    alternateIpPrimaryPort,
                    alternateIpPrimaryPort,
                    false,
                    false);
            case MAPPING_ALTERNATE_IP_AND_PORT -> new ProbeRequest(
                    generation,
                    probe,
                    otherEndpoint,
                    otherEndpoint,
                    false,
                    false);
        };
    }

    private boolean accepts(int expectedGeneration, Probe probe) {
        return expectedGeneration == generation && pendingProbe == probe && !complete;
    }

    private Transition accepted(ProbeRequest nextProbe) {
        return new Transition(true, nextProbe, snapshot());
    }

    private Transition ignored() {
        return new Transition(false, null, snapshot());
    }

    private static void requireResolved(InetSocketAddress endpoint, String name) {
        Objects.requireNonNull(endpoint, name);
        if (endpoint.isUnresolved() || endpoint.getAddress() == null || endpoint.getPort() <= 0) {
            throw new IllegalArgumentException(name + " must be a resolved UDP endpoint");
        }
    }

    private static boolean sameEndpoint(InetSocketAddress first, InetSocketAddress second) {
        return first != null
                && second != null
                && first.getPort() == second.getPort()
                && sameAddress(first, second);
    }

    private static boolean sameAddress(InetSocketAddress first, InetSocketAddress second) {
        return first != null
                && second != null
                && first.getAddress() != null
                && first.getAddress().equals(second.getAddress());
    }

    enum Probe {
        FILTER_CHANGE_IP_AND_PORT("rfc5780-filter-change-ip-port"),
        FILTER_CHANGE_PORT("rfc5780-filter-change-port"),
        MAPPING_ALTERNATE_IP("rfc5780-mapping-alternate-ip"),
        MAPPING_ALTERNATE_IP_AND_PORT("rfc5780-mapping-alternate-ip-port");

        private final String role;

        Probe(String role) {
            this.role = role;
        }

        String role() {
            return role;
        }
    }

    record ProbeRequest(int generation,
                        Probe probe,
                        InetSocketAddress targetEndpoint,
                        InetSocketAddress expectedResponseEndpoint,
                        boolean changeIp,
                        boolean changePort) {
    }

    record Snapshot(int generation,
                    String discovery,
                    String mappingBehavior,
                    String filteringBehavior,
                    InetSocketAddress mappedEndpoint,
                    boolean complete) {
    }

    record Transition(boolean accepted, ProbeRequest nextProbe, Snapshot snapshot) {
    }
}
